sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp

```diff
@@ -0,0 +1,324 @@
+//==------ invoke_simd.hpp - SYCL invoke_simd extension --*- C++ -*---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+// Implemenation of the sycl_ext_oneapi_invoke_simd extension.
+// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_invoke_simd.asciidoc
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+// SYCL extension macro definition as required by the SYCL specification.
+// 1 - Initial extension version. Base features are supported.
+#define SYCL_EXT_ONEAPI_INVOKE_SIMD 1
+
+#include <sycl/ext/oneapi/experimental/uniform.hpp>
+
+#include <CL/sycl/sub_group.hpp>
+#include <std/experimental/simd.hpp>
+#include <sycl/detail/boost/mp11.hpp>
+
+#include <functional>
+
+// TODOs:
+// * (a) TODO bool translation in spmd2simd.
+// * (b) TODO enforce constness of a functor/lambda's () operator
+// * (c) TODO support lambdas and functors in BE
+
+/// Middle End - to - Back End interface to invoke explicit SIMD functions from
+/// SPMD SYCL context. Must not be used by user code. BEs are expected to
+/// recognize this intrinsic and transform the intrinsic call with a direct call
+/// to the SIMD target, as well as process SPMD arguments in the way described
+/// in the specification for `invoke_simd`.
+/// @tparam SpmdRet the return type. Can be `uniform<T>`.
+/// @tparam SimdCallee the type of the SIMD callee function (the "target"). Must
+///  be a function type (not lambda or functor).
+/// @tparam SpmdArgs The original SPMD arguments passed to the invoke_simd.
+template <bool IsFunc, class SpmdRet, class SimdCallee, class... SpmdArgs,
+          class = std::enable_if_t<!IsFunc>>
+SYCL_EXTERNAL __regcall SpmdRet
+__builtin_invoke_simd(SimdCallee target, const void *obj, SpmdArgs... args)
+#ifdef __SYCL_DEVICE_ONLY__
+    ;
+#else
+{
+  // __builtin_invoke_simd is not supported on the host device yet
+  throw sycl::feature_not_supported();
```

Comment:
Please do not use the deprecated API. I recently made corresponding change: https://github.com/intel/llvm/pull/5861

Reply:
```suggestion
  __ESIMD_UNSUPPORTED_ON_HOST;
```

Reply:
thanks, will address in a separate PR.